### PR TITLE
f/aws_sfn_state_machine: enable tracing config

### LINKED ISSUE #12506: Chaning definition of aws_sfn_state_machine doesn't update the state machine in AWS
### Terraform Version
0.11

### Affected Resource(s)
aws_sfn_state_machine

### Expected Behavior

Modifying the "Definition" of aws_sfn_state_machine should also update the aws_sfn_state_machine in AWS.

### Actual Behavior

Modifying the "Definition" of aws_sfn_state_machine hasn't change the definition of aws_sfn_state_machine in AWS.

### Steps to Reproduce

1. Add a terraform resource like this:

```
resource "aws_sfn_state_machine" "rebuilder_v2" {
    name = "life-aggregates-service-rebuilder"
    role_arn = "${aws_iam_role.rebuilder_state_machine.arn}"

  tags {
    Environment = "${var.target_name}"
    Project = "${var.lifeomic_project}"
  }

  definition = <<EOF
{
  "Comment": "Rebuilder",
  "StartAt": "Rebuild",
  "States": {
    "Rebuild": {
      "Type": "Pass",
      "Result": [
          {"segment": 0, "totalSegments": 2, "rebuildType": "total"},
          {"segment": 1, "totalSegments": 2, "rebuildType": "total"},
      ],
      "Next": "Fanout"
    },
    "Fanout": {
      "Type": "Map",
      "Iterator": {
        "StartAt": "Scan",
        "States": {
          "Scan": {
            "Type": "Task",
            "Resource": "${ module.life_aggregates_service_rebuilder.deployment_alias_arn }",
            "InputPath": "$",
            "ResultPath": "$",
            "Retry": [
              {
                "ErrorEquals": [ "States.TaskFailed" ],
                "IntervalSeconds": 2,
                "MaxAttempts": 6,
                "BackoffRate": 4.0
              }
            ],
            "Next": "IsDone"
          },
          "IsDone": {
            "Type": "Choice",
            "Choices": [
              {
                "Variable": "$.nextToken",
                "StringEquals": "",
                "Next": "Done"
              }
            ],
            "Default": "Scan"
          },
          "Done": {
            "Type": "Pass",
            "End": true
          }
        }
      },
      "End": true
    }
  }
}
EOF
}
```

2. `terraform apply`

3. Modify aws_sfn_state_machine Definition. Specifically, I changed these lines:

```
"Result": [
          {"segment": 0, "totalSegments": 2, "rebuildType": "total"},
          {"segment": 1, "totalSegments": 2, "rebuildType": "total"}
      ],
```

To these:
```"Result": [
          {"segment": 0, "totalSegments": 3, "rebuildType": "total"},
          {"segment": 1, "totalSegments": 3, "rebuildType": "total"},
          {"segment": 2, "totalSegments": 3, "rebuildType": "total"}
      ],
```

When I ran the state machine (right after I was expecting it to update), individual lambdas received `totalSegments` of 2

### LINKED ISSUE #15430: Support X-Ray Tracing Config for Step Functions
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS recently released X-Ray tracing for Step Functions (see [here](https://aws.amazon.com/about-aws/whats-new/2020/09/aws-step-functions-adds-support-for-aws-x-ray/)). The aws_sfn_state_machine resource should provide support for this attribute.

### New or Affected Resource(s)

* aws_sfn_state_machine

### Potential Terraform Configuration

```hcl
resource "aws_sfn_state_machine" "sfn_state_machine" {
  name     = "state_machine_name"
  role_arn = "role_arn"
  definition = "definition"
  tracing_config {
    enabled = true
  }
}
```

### LINKED ISSUE #18614: tests/resource/aws_sfn_state_machine: TestAccAWSSfnStateMachine_expressLoggingConfiguration Configuration Error
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The `TestAccAWSSfnStateMachine_expressLoggingConfiguration` acceptance test consistently fails with:

```
=== CONT  TestAccAWSSfnStateMachine_expressLoggingConfiguration
resource_aws_sfn_state_machine_test.go:211: Step 1/2 error: Error running apply: exit status 1
Error: Error creating Step Function State Machine: InvalidLoggingConfiguration: Invalid Logging Configuration: Log Group ARN must be provided with '*' qualifier.
on terraform_plugin_test.tf line 105, in resource "aws_sfn_state_machine" "foo":
105: resource "aws_sfn_state_machine" "foo" {
--- FAIL: TestAccAWSSfnStateMachine_expressLoggingConfiguration (44.05s)
```

Likely requires doing exactly what the API error says to the test configuration.

### Affected Resource

* aws_sfn_state_machine

### LINKED ISSUE #18615: tests/resource/aws_sfn_state_machine: Consistent Update Test Failures
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The following acceptance testing fails consistently:

```
=== CONT  TestAccAWSSfnStateMachine_standardUpdate
resource_aws_sfn_state_machine_test.go:104: Step 2/2 error: Check failed: Check 5/7 error: aws_sfn_state_machine.foo: Attribute 'definition' didn't match ".*\\\"MaxAttempts\\\": 10.*", got "{\n  \"Comment\": \"A Hello World example of the Amazon States Language using an AWS Lambda Function with STANDARD State Machine\",\n  \"StartAt\": \"HelloWorld\",\n  \"States\": {\n    \"HelloWorld\": {\n      \"Type\": \"Task\",\n      \"Resource\": \"arn:aws:lambda:us-west-2:*******:function:sfn-x9wed6g89w\",\n      \"Retry\": [\n        {\n          \"ErrorEquals\": [\"States.ALL\"],\n          \"IntervalSeconds\": 5,\n          \"MaxAttempts\": 5,\n          \"BackoffRate\": 8.0\n        }\n      ],\n      \"End\": true\n    }\n  }\n}\n"
--- FAIL: TestAccAWSSfnStateMachine_standardUpdate (67.36s)

=== CONT  TestAccAWSSfnStateMachine_createUpdate
resource_aws_sfn_state_machine_test.go:21: Step 3/3 error: Check failed: Check 6/7 error: aws_sfn_state_machine.test: Attribute 'definition' didn't match ".*\\\"MaxAttempts\\\": 10.*", got "{\n  \"Comment\": \"A Hello World example of the Amazon States Language using an AWS Lambda Function\",\n  \"StartAt\": \"HelloWorld\",\n  \"States\": {\n    \"HelloWorld\": {\n      \"Type\": \"Task\",\n      \"Resource\": \"arn:aws:lambda:us-west-2:*******:function:tf-acc-test-6443416242939935176\",\n      \"Retry\": [\n        {\n          \"ErrorEquals\": [\n            \"States.ALL\"\n          ],\n          \"IntervalSeconds\": 5,\n          \"MaxAttempts\": 5,\n          \"BackoffRate\": 8\n        }\n      ],\n      \"End\": true\n    }\n  }\n}\n"
--- FAIL: TestAccAWSSfnStateMachine_createUpdate (84.29s)

=== CONT  TestAccAWSSfnStateMachine_expressUpdate
resource_aws_sfn_state_machine_test.go:65: Step 2/2 error: Check failed: Check 5/7 error: aws_sfn_state_machine.foo: Attribute 'definition' didn't match ".*\\\"MaxAttempts\\\": 10.*", got "{\n  \"Comment\": \"A Hello World example of the Amazon States Language using an AWS Lambda Function with EXPRESS State Machine\",\n  \"StartAt\": \"HelloWorld\",\n  \"States\": {\n    \"HelloWorld\": {\n      \"Type\": \"Task\",\n      \"Resource\": \"arn:aws:lambda:us-west-2:*******:function:sfn-fc7q9f7z80\",\n      \"Retry\": [\n        {\n          \"ErrorEquals\": [\"States.ALL\"],\n          \"IntervalSeconds\": 5,\n          \"MaxAttempts\": 5,\n          \"BackoffRate\": 8.0\n        }\n      ],\n      \"End\": true\n    }\n  }\n}\n"
--- FAIL: TestAccAWSSfnStateMachine_expressUpdate (132.07s)
```

There are a few things to check:

* Test checks match the test configuration
* The resource properly calls the API to update
* If the resource updates need an API status waiter or Terraform attribute waiter

### Affected Resource

* aws_sfn_state_machine

### References

Potentially related:

* https://github.com/hashicorp/terraform-provider-aws/issues/12506

### PR DESCRIPTION
**Pull request by nikhil-goenka**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #15430.
Closes #12506.
Closes #18614.
Closes #18615.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note

```

Output from acceptance testing:

```
make testacc TESTARGS='-run=TestAccAWSSfnStateMachine_Tracing_Config'
=== RUN   TestAccAWSSfnStateMachine_Tracing_Config
=== PAUSE TestAccAWSSfnStateMachine_Tracing_Config
=== CONT  TestAccAWSSfnStateMachine_Tracing_Config
--- PASS: TestAccAWSSfnStateMachine_Tracing_Config (168.47s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	171.227s
...
```

### CODE REVIEW

File: aws/resource_aws_sfn_state_machine_test.go [hunk 1]
```diff
@@ -359,3 +392,71 @@ EOF
 }
 `, rName, tag1Key, tag1Value, tag2Key, tag2Value)
 }
+
+func testAccAWSSfnStateMachineConfigTracingConfigEnable(rName string) string {
```
Comment (nikhil-goenka) [inline aws/resource_aws_sfn_state_machine_test.go:396]: @gdavison : is there some different command to run to format this ?

**Comment by lu-liu-rft** [comment]
Hi any update on this feature?

**Comment by lu-liu-rft** [comment]
OMG Does this mean 'X-Ray tracing enable' config will be supported for StepFunction very soon?

**Comment by nikhil-goenka** [comment]
> OMG Does this mean 'X-Ray tracing enable' config will be supported for StepFunction very soon?

As soon as it get reviewed and merged

**Comment by nikhil-goenka** [comment]
@bflad @gdavison @teamterraform 
Hi Guys,
hope you are doing well, can we please get this reviewed and merged if all looks good.

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

##### Commercial

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAWSSfnStateMachine_'                           
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSfnStateMachine_ -timeout 180m
=== RUN   TestAccAWSSfnStateMachine_createUpdate
=== PAUSE TestAccAWSSfnStateMachine_createUpdate
=== RUN   TestAccAWSSfnStateMachine_expressUpdate
=== PAUSE TestAccAWSSfnStateMachine_expressUpdate
=== RUN   TestAccAWSSfnStateMachine_standardUpdate
=== PAUSE TestAccAWSSfnStateMachine_standardUpdate
=== RUN   TestAccAWSSfnStateMachine_tags
=== PAUSE TestAccAWSSfnStateMachine_tags
=== RUN   TestAccAWSSfnStateMachine_TracingConfiguration
=== PAUSE TestAccAWSSfnStateMachine_TracingConfiguration
=== RUN   TestAccAWSSfnStateMachine_disappears
=== PAUSE TestAccAWSSfnStateMachine_disappears
=== RUN   TestAccAWSSfnStateMachine_expressLoggingConfiguration
=== PAUSE TestAccAWSSfnStateMachine_expressLoggingConfiguration
=== CONT  TestAccAWSSfnStateMachine_createUpdate
=== CONT  TestAccAWSSfnStateMachine_TracingConfiguration
=== CONT  TestAccAWSSfnStateMachine_expressLoggingConfiguration
=== CONT  TestAccAWSSfnStateMachine_standardUpdate
=== CONT  TestAccAWSSfnStateMachine_tags
=== CONT  TestAccAWSSfnStateMachine_disappears
=== CONT  TestAccAWSSfnStateMachine_expressUpdate
--- PASS: TestAccAWSSfnStateMachine_expressUpdate (82.85s)
--- PASS: TestAccAWSSfnStateMachine_standardUpdate (86.95s)
--- PASS: TestAccAWSSfnStateMachine_createUpdate (91.58s)
--- PASS: TestAccAWSSfnStateMachine_disappears (97.47s)
--- PASS: TestAccAWSSfnStateMachine_expressLoggingConfiguration (150.39s)
--- PASS: TestAccAWSSfnStateMachine_TracingConfiguration (153.24s)
--- PASS: TestAccAWSSfnStateMachine_tags (158.60s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	161.891s
```

##### GovCloud

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAWSSfnStateMachine_'   
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSfnStateMachine_ -timeout 180m
=== RUN   TestAccAWSSfnStateMachine_createUpdate
=== PAUSE TestAccAWSSfnStateMachine_createUpdate
=== RUN   TestAccAWSSfnStateMachine_expressUpdate
=== PAUSE TestAccAWSSfnStateMachine_expressUpdate
=== RUN   TestAccAWSSfnStateMachine_standardUpdate
=== PAUSE TestAccAWSSfnStateMachine_standardUpdate
=== RUN   TestAccAWSSfnStateMachine_tags
=== PAUSE TestAccAWSSfnStateMachine_tags
=== RUN   TestAccAWSSfnStateMachine_TracingConfiguration
=== PAUSE TestAccAWSSfnStateMachine_TracingConfiguration
=== RUN   TestAccAWSSfnStateMachine_disappears
=== PAUSE TestAccAWSSfnStateMachine_disappears
=== RUN   TestAccAWSSfnStateMachine_expressLoggingConfiguration
=== PAUSE TestAccAWSSfnStateMachine_expressLoggingConfiguration
=== CONT  TestAccAWSSfnStateMachine_createUpdate
=== CONT  TestAccAWSSfnStateMachine_TracingConfiguration
=== CONT  TestAccAWSSfnStateMachine_standardUpdate
=== CONT  TestAccAWSSfnStateMachine_expressLoggingConfiguration
=== CONT  TestAccAWSSfnStateMachine_tags
=== CONT  TestAccAWSSfnStateMachine_disappears
=== CONT  TestAccAWSSfnStateMachine_expressUpdate
--- PASS: TestAccAWSSfnStateMachine_disappears (59.21s)
--- PASS: TestAccAWSSfnStateMachine_standardUpdate (107.93s)
--- PASS: TestAccAWSSfnStateMachine_tags (109.99s)
--- PASS: TestAccAWSSfnStateMachine_createUpdate (111.59s)
--- PASS: TestAccAWSSfnStateMachine_expressLoggingConfiguration (111.89s)
--- PASS: TestAccAWSSfnStateMachine_expressUpdate (114.22s)
--- PASS: TestAccAWSSfnStateMachine_TracingConfiguration (117.20s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	120.135s
```

**Comment by ewbankkit** [comment]
@nikhil-goenka Thanks for the contribution 👏.
I took the liberty of tweaking the configuration to be `tracing_configuration` (and fixed the acceptance tests) in order to get this one merged quicker.

**Pull request merged**